### DISCUSSION TITLE
Creating a Grammar for QB64/QBasic

### DISCUSSION THREAD

**Discussion by dazlab** [pulsar-edit/.github#163] [Community Packages]
Bit of an oddball request, but how would one go about creating a grammar for QB64 - which is a modern extension of QuickBASIC? Can it be done without having to manually code a parser, or would it require a fair bit of coding?

**Comment by Daeraxa** [comment]
We have docs on how to make grammars depending on what type of grammar you want to make.
Tree-sitter: https://pulsar-edit.dev/docs/launch-manual/sections/core-hacking/#creating-a-grammar
Legacy textmate: https://pulsar-edit.dev/docs/launch-manual/sections/core-hacking/#creating-a-legacy-textmate-grammar

However we are currently in the process of migrating the somewhat unique (and orphaned) Atom tree-sitter implementation to the more standardised WASM tree-sitter which has different requirements so you may wish to wait for that to become available with updated docs.

**Reply by confused-Techie** [reply]
Probably the easiest thing to do would be to create a legacy TextMate Grammar. It really doesn't take all that much coding, but is more of just figuring out the right way to express the structure of the language.

But in reality only needs a few files to do so.

The other thing I would suggest is see if you can find a [TextMate Bundle](https://packagecontrol.io/search/qb64) (Linked the search for one on a site, but I couldn't locate anything there, but you may have luck with other keywords) for the language, the docs linked above should also have a section on using PPM to automatically convert a TextMate Bundle to being a Pulsar Compatible TextMate Grammar.

Otherwise something you could try to at least get a head start is looking at the languages supported by GitHub's [Linguist](https://github.com/github/linguist/tree/master/vendor#readme). Since commonly, many of the grammars used here will either be Pulsar TextMate Grammars, or TextMate Bundles, or Tree-Sitter Grammars, which all can be used to work with Pulsar with minimal effort. But unfortunately I didn't have much luck finding this particular language. But hopefully these links can lead you in the right place.

**Comment by confused-Techie** [comment]
@dazlab If you'd like I'd be more than happy to provide additional resources, or even help getting a jumping off point for a grammar. Or otherwise take a shot at it myself. Just let me know!

**Reply by dazlab** [reply]
Apologies, been caught up in other projects. 

I followed your TextMate suggestion. I know it's considered legacy and everything is migrating to tree-sitter, but it's easier to get started with. Is it just a case of packaging up the grammar file? Here's what I have:

`{
	"scopeName": "source.qbasic",
	"uuid": "e7de52b7-8378-4396-848e-e9da04e24ddb",
	"fileTypes": [
		"bas",
		"qb",
		"qbasic"
	],
	"name": "QBasic",
	"patterns": [
		{
			"captures": {
				"1": {
					"name": "punctuation.definition.comment.qbasic"
				}
			},
			"comment": "Comment",
			"match": "(').*",
			"name": "comment.line.qbasic"
		},
		{
			"comment": "Delimiter",
			"match": "[,:;]",
			"name": "meta.delimiter.object.qbasic"
		},
		{
			"comment": "Keyword",
			"match": "(?i)(\\b((END )?IF|(END )?SELECT|(RESUME )?NEXT|CASE|CLOSE|DO|ELSE|FOR|GOSUB|GOTO|LOOP|ON|OPEN|RETURN|THEN|TO|UNTIL|WHILE)\\b)",
			"name": "keyword.control.qbasic"
		},
		{
			"comment": "Function",
			"match": "(?i)(\\b(CHR$|CLS|COLOR|DATA|INPUT|INT|KEY|LEN|LINE|LOCATE|PALETTE|PLAY|PRINT|PSET|PUT|REDIM|RND|SCREEN|TAB|VIEW)\\b)",
			"name": "support.function.qbasic"
		},
		{
			"comment": "Operator",
			"match": "(?i)((\\+|=|<|>|    <>|AND|OR))",
			"name": "keyword.operator.qbasic"
		},
		{
			"comment": "Numeric",
			"match": "\\b(\\d(\\.\\d)?)+",
			"name": "constant.numeric.qbasic"
		},
		{
			"comment": "Global",
			"match": "\\b\\w+\\$",
			"name": "constant.global.qbasic"
		},
		{
			"captures": {
				"1": {
					"name": "entity.name.function.qbasic"
				}
			},
			"comment": "SUB",
			"match": "(?i)(^(\\w+):)",
			"name": "meta.function.qbasic"
		},
		{
			"comment": "Brace, round",
			"match": "[\\(\\)]",
			"name": "meta.brace.round.qbasic"
		},
		{
			"comment": "Brace, curly",
			"match": "[\\{\\}]",
			"name": "meta.brace.curly.qbasic"
		},
		{
			"begin": "(\\w+\\()",
			"beginCaptures": {
				"1": {
					"name": "constant.array.qbasic"
				}
			},
			"comment": "Array",
			"end": "(\\))",
			"endCaptures": {
				"1": {
					"name": "constant.array.qbasic"
				}
			},
			"name": "other.array.qbasic",
			"patterns": [
				{
					"include": "$self"
				}
			]
		},
		{
			"begin": "(\")",
			"beginCaptures": {
				"1": {
					"name": "punctuation.definition.string.begin.qbasic"
				}
			},
			"comment": "String, double-quoted",
			"end": "(\")",
			"endCaptures": {
				"1": {
					"name": "punctuation.definition.string.begin.qbasic"
				}
			},
			"name": "string.quoted.double.qbasic",
			"patterns": [
				{
					"comment": "Escaped double-quote inside double-quoted string",
					"match": "(\\\")",
					"name": "constant.character.escape.qbasic"
				},
				{
					"comment": "Single quote inside double-quoted string",
					"match": "(')",
					"name": "other.qbasic"
				},
				{
					"include": "$self"
				}
			]
		}
	]
}`

**Reply by confused-Techie** [reply]
@dazlab From a quick glance this does in fact look like a perfectly valid grammar file, even made sure to get double escapes for nearly everything, (Except "comment" I believe should be `"(\\")"`)

But yeah from there just the bear bones of a package such as `package.json` and a folder named `grammars` then place what you have into a `language-qbasic.cson` file and you are good to go!

But yeah I don't blame you for going the TextMate route, it's honestly a much simpler process **depending on the language**, so in some cases it's absolutely the right move.

But if you'd like some new examples of how to package up your grammar here's a couple I've recently put together that can maybe serve a reference point:

* [`langauge-robots-txt`](https://github.com/confused-Techie/language-robots-txt)
* [`language-pon`](https://github.com/confused-Techie/language-pon)

**Reply by dazlab** [reply]
I managed to get packaged and it is now working! Awesome. Thank you so much for your assistance.

**Reply by confused-Techie** [reply]
Of course @dazlab, feel free to shoot me a ping when/if you have it published, would love to know about it
